### DISCUSSION TITLE
ESP32 internal temperature is not correctly updated with teleperiod

### DISCUSSION THREAD

**Discussion by Bebuccio** [arendst/Tasmota#20505] [Q&A]
I have found that the Internal Temperature (CPU-Temp) of ESP32 is non correctly updated when reported by SENSOR message (teleperiod).
Enabling the watch of internal temperature (SetOption146 1) and forcing the teleperiod every 20 seconds (teleperiod 20) I see that 
the temperature sent by sensor message is often of 53,3 degree also if the true temperature is 55/58 degree (I see these values on the main WEB page).

16:18:25.696 RSL: STATE = {"Time":"2024-01-15T16:18:25","Uptime":"0T00:29:01","UptimeSec":1741,"Heap":136,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":16,"Objects":287},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":50,"Signal":-75,"LinkCount":1,"Downtime":"0T00:00:04"}}
16:18:25.705 RSL: SENSOR = {"Time":"2024-01-15T16:18:25","ESP32":{"Temperature":53.3},"TempUnit":"C"}
16:18:45.723 RSL: STATE = {"Time":"2024-01-15T16:18:45","Uptime":"0T00:29:21","UptimeSec":1761,"Heap":136,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":16,"Objects":287},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":50,"Signal":-75,"LinkCount":1,"Downtime":"0T00:00:04"}}
16:18:45.732 RSL: SENSOR = {"Time":"2024-01-15T16:18:45","ESP32":{"Temperature":53.3},"TempUnit":"C"}
16:19:05.725 RSL: STATE = {"Time":"2024-01-15T16:19:05","Uptime":"0T00:29:41","UptimeSec":1781,"Heap":133,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":16,"Objects":287},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":48,"Signal":-76,"LinkCount":1,"Downtime":"0T00:00:04"}}
16:19:05.734 RSL: SENSOR = {"Time":"2024-01-15T16:19:05","ESP32":{"Temperature":56.1},"TempUnit":"C"}
16:19:25.708 RSL: STATE = {"Time":"2024-01-15T16:19:25","Uptime":"0T00:30:01","UptimeSec":1801,"Heap":136,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":16,"Objects":287},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":48,"Signal":-76,"LinkCount":1,"Downtime":"0T00:00:04"}}
16:19:25.717 RSL: SENSOR = {"Time":"2024-01-15T16:19:25","ESP32":{"Temperature":53.3},"TempUnit":"C"}
16:19:45.737 RSL: STATE = {"Time":"2024-01-15T16:19:45","Uptime":"0T00:30:21","UptimeSec":1821,"Heap":136,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":16,"Objects":287},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":48,"Signal":-76,"LinkCount":1,"Downtime":"0T00:00:04"}}
16:19:45.745 RSL: SENSOR = {"Time":"2024-01-15T16:19:45","ESP32":{"Temperature":53.3},"TempUnit":"C"}
16:20:05.747 RSL: STATE = {"Time":"2024-01-15T16:20:05","Uptime":"0T00:30:41","UptimeSec":1841,"Heap":136,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":16,"Objects":287},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":50,"Signal":-75,"LinkCount":1,"Downtime":"0T00:00:04"}}
16:20:05.756 RSL: SENSOR = {"Time":"2024-01-15T16:20:05","ESP32":{"Temperature":53.3},"TempUnit":"C"}
16:20:25.723 RSL: STATE = {"Time":"2024-01-15T16:20:25","Uptime":"0T00:31:01","UptimeSec":1861,"Heap":136,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":16,"Objects":287},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":48,"Signal":-76,"LinkCount":1,"Downtime":"0T00:00:04"}}
16:20:25.732 RSL: SENSOR = {"Time":"2024-01-15T16:20:25","ESP32":{"Temperature":53.3},"TempUnit":"C"}
16:20:45.740 RSL: STATE = {"Time":"2024-01-15T16:20:45","Uptime":"0T00:31:21","UptimeSec":1881,"Heap":136,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":16,"Objects":287},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":48,"Signal":-76,"LinkCount":1,"Downtime":"0T00:00:04"}}
16:20:45.749 RSL: SENSOR = {"Time":"2024-01-15T16:20:45","ESP32":{"Temperature":53.3},"TempUnit":"C"}
16:21:05.708 RSL: STATE = {"Time":"2024-01-15T16:21:05","Uptime":"0T00:31:41","UptimeSec":1901,"Heap":136,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":16,"Objects":287},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":46,"Signal":-77,"LinkCount":1,"Downtime":"0T00:00:04"}}
16:21:05.717 RSL: SENSOR = {"Time":"2024-01-15T16:21:05","ESP32":{"Temperature":53.3},"TempUnit":"C"}
16:21:25.706 RSL: STATE = {"Time":"2024-01-15T16:21:25","Uptime":"0T00:32:01","UptimeSec":1921,"Heap":136,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":16,"Objects":287},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":48,"Signal":-76,"LinkCount":1,"Downtime":"0T00:00:04"}}
16:21:25.715 RSL: SENSOR = {"Time":"2024-01-15T16:21:25","ESP32":{"Temperature":56.1},"TempUnit":"C"}

This problem is present also in the last Tasmota version (13.3.0) and this was reflected to matter-data if this feature-protocol is enabled

**Comment by sfromis** [comment]
Doing a small test, I'm seeing perfect alignment between values in the JSON and on the Tasmota front page, within the 0.06 °C rounding of the values (I'm using `TempRes 2`)

Anyway, the reason why you need to use an option to enable the value is that has been found to be very bad, not with a clear relaitionship to a "reasonable" temperature, like when I've seen it indicate near freezing in a room with normal ambient indoor air temperature. Basically the word "true" just does not apply.

Are you using a `TempOffset`? Not that this should make a difference, as it should be applied both places.

**Comment by Bebuccio** [comment]
This my configuration of tempres and tempoffset:

17:07:02.702 CMD: tempoffset
17:07:02.706 RSL: RESULT = {"TempOffset":0.0}
17:07:08.700 CMD: tempres
17:07:08.704 RSL: RESULT = {"TempRes":1

If I use the status 8 command to retrieve the temperature, the value is very often correct (some times is uncorrect).
Indeed waiting for a teleperiod to retrieve the temperature, this is very often **UNCORRECT**
Forcing a teleperiod low (10 secs) you can see the wrong values mixed with the teleperiod values

17:14:15.043 RSL: STATUS8 = {"StatusSNS":{"Time":"2024-01-15T17:14:15","ESP32":{"Temperature":59.44},"TempUnit":"C"}}
17:14:15.450 CMD: status 8
17:14:15.459 RSL: STATUS8 = {"StatusSNS":{"Time":"2024-01-15T17:14:15","ESP32":{"Temperature":59.44},"TempUnit":"C"}}
17:14:15.768 CMD: status 8
17:14:15.777 RSL: STATUS8 = {"StatusSNS":{"Time":"2024-01-15T17:14:15","ESP32":{"Temperature":59.44},"TempUnit":"C"}}
17:14:15.990 CMD: status 8
17:14:16.000 RSL: STATUS8 = {"StatusSNS":{"Time":"2024-01-15T17:14:15","ESP32":{"Temperature":60.00},"TempUnit":"C"}}
17:14:16.314 CMD: status 8
17:14:16.325 RSL: STATUS8 = {"StatusSNS":{"Time":"2024-01-15T17:14:16","ESP32":{"Temperature":**53.33**},"TempUnit":"C"}}
17:14:16.685 CMD: status 8
17:14:16.694 RSL: STATUS8 = {"StatusSNS":{"Time":"2024-01-15T17:14:16","ESP32":{"Temperature":60.00},"TempUnit":"C"}}
17:14:16.977 CMD: status 8
17:14:16.987 RSL: STATUS8 = {"StatusSNS":{"Time":"2024-01-15T17:14:16","ESP32":{"Temperature":60.00},"TempUnit":"C"}}
17:14:17.155 RSL: STATE = {"Time":"2024-01-15T17:14:17","Uptime":"0T00:19:38","UptimeSec":1178,"Heap":84,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":61,"MqttCount":0,"Berry":{"HeapUsed":48,"Objects":670},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":48,"Signal":-76,"LinkCount":1,"Downtime":"0T00:00:03"}}
17:14:17.171 RSL: SENSOR = {"Time":"2024-01-15T17:14:17","ESP32":{"Temperature":**53.33**},"TempUnit":"C"}
17:14:17.297 CMD: status 8
17:14:17.306 RSL: STATUS8 = {"StatusSNS":{"Time":"2024-01-15T17:14:17","ESP32":{"Temperature":60.00},"TempUnit":"C"}}
17:14:17.574 CMD: status 8
17:14:17.583 RSL: STATUS8 = {"StatusSNS":{"Time":"2024-01-15T17:14:17","ESP32":{"Temperature":60.00},"TempUnit":"C"}}
17:14:27.116 RSL: STATE = {"Time":"2024-01-15T17:14:27","Uptime":"0T00:19:48","UptimeSec":1188,"Heap":80,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":38,"MqttCount":0,"Berry":{"HeapUsed":48,"Objects":667},"Wifi":{"AP":1,"SSId":"Dxxxxxxxt","BSSId":"4yyyyyyyyy9","Channel":5,"Mode":"11n","RSSI":46,"Signal":-77,"LinkCount":1,"Downtime":"0T00:00:03"}}
17:14:27.132 RSL: SENSOR = {"Time":"2024-01-15T17:14:27","ESP32":{"Temperature":**53.33**},"TempUnit":"C"}

**Reply by sfromis** [reply]
Since you see similarly varying values using the command `Status 8`, this looks like it is the already mentioned **bad**ness of the ESP32 internal temperature sensing, just not "correct" for *anything*.

In my experience, that value is so dubious that I find it better to just forget about it. Of course, this also means that I've not even worried about whether the badness could be partially mitigated from Tasmota.

**Comment by Wintespe** [comment]
https://docs.espressif.com/projects/esp-idf/en/latest/esp32c6/api-reference/peripherals/temp_sensor.html

The temperature sensor is designed primarily to measure the temperature changes inside the chip. The internal temperature of a chip is usually higher than the ambient temperature, and is affected by factors such as the microcontroller's clock frequency or I/O load, and the external thermal environment.

**Reply by Bebuccio** [reply]
Hi Wintespe.
I know that the CPU-Temp is not the room-temp.
The problem is that this value sent via Status 8 or Teleperiod is not always the current value of the sensor. The value shown by main web page is ok: when I ask via status 8 or sent via periodic teleperiod some times is the same of the main web page and other times is a fixed value 53.3
The routine that "picks" the cpu-temp value with the status 8 command or within teleperiod some times get a value of 53.3 degree.

**Reply by Wintespe** [reply]
If the value in the web main page is plausible and only the temperature value of status 8 shows the problem, I suspect that there is a problem with Tasmota. I hope the Tasmota team can help you.
